kernel/kernel-impl/src/main/java/org/sakaiproject/db/impl/BasicSqlService.java

```diff
@@ -1531,7 +1531,7 @@ else if (recordAlreadyExists)
 			//make sure we return the connection even if the rollback etc above
 			// KNL-492 connFromThreadLocal is tested so we can avoid returning a
 			// connection that is being managed elsewhere
-			if (conn != null && !connFromThreadLocal)
+			if (conn != null && connFromThreadLocal)
```

Comment:
Are you sure about this change? If the connection is managed in threadlocal, then it shouldn't be returned to the pool, right?

Reply:
Well, can anyone be sure after reading the code in BasicSqlService... It feels like at least two classes/use cases in the same class.
The comment says it all: Do not return a connection that is outer scope, so I think the code should do what the comment says.

When using my correction you do not have to get a connection in the ClusterEventTracking, which seems just wrong as we already have a connection.

It all depends on how the class is used but from my investigation the are not many uses. If the dbInsert() is used without connection it will borrow and then return, and it will never be placed in the threadLocal.

Reply:
Hi @jayee 
The connection managed in ThreadLocal should not be returned here that is what the code explicitly says and the comments (KNL-492) say the same.

Reply:
@adrianfish is this your code from KNL-492?

Reply:
"connection that is being managed elsewhere" - I'm interpreting that as a connection that is input to the method. Even if you do not read the comment it seems strange to alter the connection if it is someone else handling it to you.

If you don't change the BasicSqlService you can't change ClusterEventTracking the way I did.

Reply:
"connection that is being managed elsewhere" -- I don't think it's a connection that is input to the method. It appears to be a connection that is put into ThreadLocal

Reply:
So essentially what "connection that is being managed elsewhere" aka ThreadLocal is that there is an ongoing transaction as a result any queries that execute in that context (like dbInsert) are part of the transaction. When the transaction is complete commit is called and the connection in ThreadLocal is cleaned up, all of this is handled internally by BasicSqlService.

I can see that possibly there is an issue with this method org.sakaiproject.event.impl.ClusterEventTracking#writeEvent if your setting
`setBatchWrite@org.sakaiproject.event.impl.ClusterEventTracking=false`? The writeEvent method should be setting up a connection before calling dbInsert or dbWrite. Anytime you call the BasicSqlService methods with no connection it will attempt to participate in the ongoing transaction which should only be used if your inside a Runnable that is going to be executed with `org.sakaiproject.db.impl.BasicSqlService#transact`. Can you confirm this?

Reply:
so @jayee I looked through the code, and I think you found something important. This line should *not* be doing a returnConnection(conn) if the conn was sent to dbInsert with a callerConn.

So how about this change instead:

if (conn != null && callerConnection == null && !connFromThreadLocal)